node/pkg/common/queryRequest.go

```diff
@@ -50,233 +98,350 @@ func PostSignedQueryRequest(signedQueryReqSendC chan<- *gossipv1.SignedQueryRequ
+func (queryRequest *QueryRequest) Marshal() ([]byte, error) {
+	if err := queryRequest.Validate(); err != nil {
+		return nil, err
+	}
+
 	buf := new(bytes.Buffer)
 
 	vaa.MustWrite(buf, binary.BigEndian, queryRequest.Nonce) // uint32
 
 	vaa.MustWrite(buf, binary.BigEndian, uint8(len(queryRequest.PerChainQueries)))
 	for _, perChainQuery := range queryRequest.PerChainQueries {
-		pcqBuf, err := MarshalPerChainQueryRequest(perChainQuery)
+		pcqBuf, err := perChainQuery.Marshal()
 		if err != nil {
-			return nil, fmt.Errorf("failed to marshal per chain query")
+			return nil, fmt.Errorf("failed to marshal per chain query: %w", err)
 		}
 		buf.Write(pcqBuf)
 	}
 
 	return buf.Bytes(), nil
 }
 
-// MarshalQueryRequest serializes the binary representation of a per chain query request
-func MarshalPerChainQueryRequest(perChainQuery *gossipv1.PerChainQueryRequest) ([]byte, error) {
-	buf := new(bytes.Buffer)
-	switch req := perChainQuery.Message.(type) {
-	case *gossipv1.PerChainQueryRequest_EthCallQueryRequest:
-		vaa.MustWrite(buf, binary.BigEndian, QUERY_REQUEST_TYPE_ETH_CALL)
-		vaa.MustWrite(buf, binary.BigEndian, uint16(perChainQuery.ChainId))
-		vaa.MustWrite(buf, binary.BigEndian, uint32(len(req.EthCallQueryRequest.Block)))
-		buf.Write([]byte(req.EthCallQueryRequest.Block))
-		vaa.MustWrite(buf, binary.BigEndian, uint8(len(req.EthCallQueryRequest.CallData)))
-		for _, callData := range req.EthCallQueryRequest.CallData {
-			buf.Write(callData.To)
-			vaa.MustWrite(buf, binary.BigEndian, uint32(len(callData.Data)))
-			buf.Write(callData.Data)
+// Unmarshal deserializes the binary representation of a query request from a byte array
+func (queryRequest *QueryRequest) Unmarshal(data []byte) error {
+	reader := bytes.NewReader(data[:])
+	return queryRequest.UnmarshalFromReader(reader)
+}
+
+// UnmarshalFromReader deserializes the binary representation of a query request from an existing reader
+func (queryRequest *QueryRequest) UnmarshalFromReader(reader *bytes.Reader) error {
+	if err := binary.Read(reader, binary.BigEndian, &queryRequest.Nonce); err != nil {
+		return fmt.Errorf("failed to read request nonce: %w", err)
+	}
+
+	numPerChainQueries := uint8(0)
+	if err := binary.Read(reader, binary.BigEndian, &numPerChainQueries); err != nil {
+		return fmt.Errorf("failed to read number of per chain queries: %w", err)
+	}
+
+	for count := 0; count < int(numPerChainQueries); count++ {
+		perChainQuery := PerChainQueryRequest{}
+		err := perChainQuery.UnmarshalFromReader(reader)
+		if err != nil {
+			return fmt.Errorf("failed to Unmarshal per chain query: %w", err)
 		}
-	default:
-		return nil, fmt.Errorf("invalid request type")
+		queryRequest.PerChainQueries = append(queryRequest.PerChainQueries, &perChainQuery)
 	}
+
+	return nil
+}
+
+// Validate does basic validation on a received query request.
+func (queryRequest *QueryRequest) Validate() error {
+	// Nothing to validate on the Nonce.
+	if len(queryRequest.PerChainQueries) == 0 {
```

Comment:
This is a Go question, I guess.  len returns an int instead of a uint.  In general, should we check for <= 0 or == 0 is sufficient?

Reply:
Seems like something lint might complain about some day. I changed it.